pkg/appmgmt/general/podevent_handler.go

```diff
@@ -0,0 +1,176 @@
+/*
+ Licensed to the Apache Software Foundation (ASF) under one
+ or more contributor license agreements.  See the NOTICE file
+ distributed with this work for additional information
+ regarding copyright ownership.  The ASF licenses this file
+ to you under the Apache License, Version 2.0 (the
+ "License"); you may not use this file except in compliance
+ with the License.  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package general
+
+import (
+	"sync"
+
+	"github.com/apache/yunikorn-k8shim/pkg/appmgmt/interfaces"
+	"github.com/apache/yunikorn-k8shim/pkg/log"
+
+	"go.uber.org/zap"
+
+	v1 "k8s.io/api/core/v1"
+)
+
+type PodEventHandler struct {
+	recoveryRunning bool
+	amProtocol      interfaces.ApplicationManagementProtocol
+	asyncEvents     []*podAsyncEvent
+	sync.Mutex
+}
+
+const (
+	AddPod = iota
+	UpdatePod
+	DeletePod
+)
+
+const (
+	Recovery = iota
+	Informers
+)
+
+type EventType int
+type EventSource int
+
+type podAsyncEvent struct {
+	eventType EventType
+	pod       *v1.Pod
+}
+
+func (p *PodEventHandler) HandleEvent(eventType EventType, source EventSource, pod *v1.Pod) interfaces.ManagedApp {
```

Comment:
Can we extract everything within the locks into a separate method that does locking using defer? that method could return a bool indicating whether the event has been processed. Something like:

```
func (p *PodEventHandler) HandleEvent(eventType EventType, source EventSource, pod *v1.Pod) interfaces.ManagedApp {
	if p.handleRecoveryEvent(eventType, source, pod) {
		return nil
	}
	return p.internalHandle(eventType, source, pod)
}


func (p *PodEventHandler) handleRecoveryEvent(eventType EventType, source EventSource, pod *v1.Pod) bool {
	p.Lock()
	defer p.Unlock()

	if p.recoveryRunning && source == Informers {
		log.Logger().Debug("Storing async event", zap.Int("eventType", int(eventType)),
			zap.String("pod", pod.GetName()))
		p.asyncEvents = append(p.asyncEvents, &podAsyncEvent{eventType, pod})
		return true
	}
	return false
}
```

Reply:
Done

Reply:
Done